### PR TITLE
[PL-22097]: review git full sync APIs

### PR DESCRIPTION
**Pull request by mohitg0795**
You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/30411)

### CODE REVIEW

**Review by ankit-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30411)* status: 5 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java, line 74 at r1](https://reviewable.io/reviews/harness/harness-core/30411#-Mup5KPy7NY8_fSjtVX0:-Mup5KPy7NY8_fSjtVX1:b-muah0a) ([raw file](https://github.com/harness/harness-core/blob/d7addc89f7392339729a3da2d2ba1207d3511de1/136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java#L74)):*
> ```Java
> import retrofit2.http.Body;
> 
> @Api("/full-sync")
> ```

git-full-sync?
___
*[136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java, line 110 at r1](https://reviewable.io/reviews/harness/harness-core/30411#-Mup5WjhDxGEDHdrsSH9:-Mup5WjhDxGEDHdrsSHA:b-gfs4si) ([raw file](https://github.com/harness/harness-core/blob/d7addc89f7392339729a3da2d2ba1207d3511de1/136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java#L110)):*
> ```Java
>       responses =
>       { @io.swagger.v3.oas.annotations.responses.ApiResponse(description = "Successfully triggered Full Sync") })
>   public ResponseDTO<TriggerFullSyncResponseDTO>
> ```

If full sync could not be triggered, what will we do? TriggerFullSyncResponseDTO would have "false" in it. But, it won't tell to user why full sync couldn't be triggered 
___
*[136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java, line 146 at r1](https://reviewable.io/reviews/harness/harness-core/30411#-Mup5p9e0298C9CpLkYA:-Mup5p9e0298C9CpLkYB:b-drmv5f) ([raw file](https://github.com/harness/harness-core/blob/d7addc89f7392339729a3da2d2ba1207d3511de1/136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java#L146)):*
> ```Java
>       @Parameter(description = PROJECT_PARAM_MESSAGE) @QueryParam(NGCommonEntityConstants.PROJECT_KEY) @io.
>       harness.accesscontrol.ProjectIdentifier @ProjectIdentifier String projectIdentifier,
>       @RequestBody(description = "Configuration body") @NotNull @Valid GitFullSyncConfigRequestDTO requestDTO) {
> ```

Wrong description. It should be more descriptive and should not use the term body
___
*[136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java, line 246 at r1](https://reviewable.io/reviews/harness/harness-core/30411#-Mup6n9fAPK1GQG6Hnd0:-Mup6n9fAPK1GQG6Hnd1:b-gr3361) ([raw file](https://github.com/harness/harness-core/blob/d7addc89f7392339729a3da2d2ba1207d3511de1/136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java#L246)):*
> ```Java
>   @NGAccessControlCheck(resourceType = ResourceTypes.PROJECT, permission = VIEW_PROJECT_PERMISSION)
>   public ResponseDTO<PageResponse<GitFullSyncEntityInfoDTO>>
>   listFiles(
> ```

API should be getFullSyncStatus
___
*[136-git-sync-manager/src/main/java/io/harness/gitsync/fullsync/dtos/GitFullSyncConfigDTO.java, line 40 at r1](https://reviewable.io/reviews/harness/harness-core/30411#-Mup4ioc-pLJzl0QfwFv:-Mup3rwh8zt9E1xdLNyr:bgh8tt8) ([raw file](https://github.com/harness/harness-core/blob/d7addc89f7392339729a3da2d2ba1207d3511de1/136-git-sync-manager/src/main/java/io/harness/gitsync/fullsync/dtos/GitFullSyncConfigDTO.java#L40)):*
> ```Java
>   // Why is this field required in response?
>   @Schema(description = "Checks the new Branch") boolean isNewBranch;
>   @Schema(description = "Target Branch for pull request") String targetBranch;
> ```

Whats the difference between targetBranch and baseBranch?
___
*[136-git-sync-manager/src/main/java/io/harness/gitsync/fullsync/dtos/GitFullSyncConfigRequestDTO.java, line 28 at r1](https://reviewable.io/reviews/harness/harness-core/30411#-Mup4U8B3RGWpP1AYPgr:-Mup4U8B3RGWpP1AYPgs:b59rdax) ([raw file](https://github.com/harness/harness-core/blob/d7addc89f7392339729a3da2d2ba1207d3511de1/136-git-sync-manager/src/main/java/io/harness/gitsync/fullsync/dtos/GitFullSyncConfigRequestDTO.java#L28)):*
> ```Java
>   @Schema(description = "Branch on which Entities will be pushed") @NotNull String branch;
>   @Schema(description = GitSyncApiConstants.REPOID_PARAM_MESSAGE) @NotNull String repoIdentifier;
>   @Schema(description = "Root Folder Path where entities will be pushed") @NotNull String rootFolder;
> ```

What is Root Folder Path? We should clarify that in the description



_Code quote:_
```Java
Root Folder Path where entities will be pushed
```
___
*[136-git-sync-manager/src/main/java/io/harness/gitsync/fullsync/dtos/GitFullSyncEntityInfoFilterDTO.java, line 38 at r1](https://reviewable.io/reviews/harness/harness-core/30411#-Mup4rhv3YN9QfEbXgBv:-Mup4rhv3YN9QfEbXgBw:b-kz8irs) ([raw file](https://github.com/harness/harness-core/blob/d7addc89f7392339729a3da2d2ba1207d3511de1/136-git-sync-manager/src/main/java/io/harness/gitsync/fullsync/dtos/GitFullSyncEntityInfoFilterDTO.java#L38)):*
> ```Java
> @NoArgsConstructor
> @AllArgsConstructor
> @Schema(name = "GitFullSyncEntityInfoFilter", description = "This contains filters for Git Full Sync")
> ```

Description is incorrect

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/fullsync/dtos/GitFullSyncConfigRequestDTO.java [hunk 1]
```diff
@@ -18,12 +18,16 @@
 
 @Data
 @Builder
+// description should be : "This defines details required to trigger full sync" ?
 @Schema(name = "GitFullSyncConfigRequest", description = "This contains details to trigger Full Sync")
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/fullsync/dtos/GitFullSyncConfigRequestDTO.java:22]: Details required to trigger full sync.

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/fullsync/dtos/GitFullSyncConfigDTO.java [hunk 2]
```diff
@@ -22,12 +27,17 @@
   @Schema(description = NGCommonEntityConstants.ORG_PARAM_MESSAGE) private String orgIdentifier;
   @Schema(description = NGCommonEntityConstants.PROJECT_PARAM_MESSAGE) private String projectIdentifier;
   @Schema(description = GitSyncApiConstants.DEFAULT_BRANCH_PARAM_MESSAGE) private String baseBranch;
+  // Entities should start with small letter
   @Schema(description = "Branch on which Entities will be pushed") private String branch;
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/fullsync/dtos/GitFullSyncConfigDTO.java:31]: Branch to which entities will be pushed.

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/fullsync/dtos/GitFullSyncConfigDTO.java [hunk 3]
```diff
@@ -22,12 +27,17 @@
...
   @Schema(description = "PR Title") private String prTitle;
+  // Why are we specifying default value in a response dto?
   @Schema(description = "This checks whether to create a pull request. Its default value is False")
   private boolean createPullRequest;
   @Schema(description = GitSyncApiConstants.REPOID_PARAM_MESSAGE) private String repoIdentifier;
+  // Wrong way of description for a response
+  // Why is this field required in response?
   @Schema(description = "Checks the new Branch") boolean isNewBranch;
   @Schema(description = "Target Branch for pull request") String targetBranch;
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/fullsync/dtos/GitFullSyncConfigDTO.java:40]: Target branch for the pull request.

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/fullsync/dtos/GitFullSyncConfigDTO.java [hunk 4]
```diff
@@ -22,12 +27,17 @@
...
+  // Entities should start with small letter
   @Schema(description = "Branch on which Entities will be pushed") private String branch;
   @Schema(description = "PR Title") private String prTitle;
+  // Why are we specifying default value in a response dto?
   @Schema(description = "This checks whether to create a pull request. Its default value is False")
   private boolean createPullRequest;
   @Schema(description = GitSyncApiConstants.REPOID_PARAM_MESSAGE) private String repoIdentifier;
+  // Wrong way of description for a response
+  // Why is this field required in response?
   @Schema(description = "Checks the new Branch") boolean isNewBranch;
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/fullsync/dtos/GitFullSyncConfigDTO.java:39]: Determines if it is a new branch.

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/fullsync/dtos/GitFullSyncConfigDTO.java [hunk 5]
```diff
@@ -14,6 +14,11 @@
...
 import lombok.Data;
 
+// This class is being used as response dto for createFullSyncConfig request, comments are accordingly given
+
+// I guess it is a common struct used for both request and response, which should never be the case.
+// Currently, all descriptions seem off as they can't align with both request and response at same time
+// Need to discuss, review and fix it
 @Data
 @Builder
 @Schema(name = "GitFullSyncConfig", description = "This has config details specific to Git Full Sync with Harness")
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/fullsync/dtos/GitFullSyncConfigDTO.java:24]: Can we say "Details to configure Git full sync with Harness"?
@mohitg0795

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/core/fullsync/GitFullSyncConfigServiceImpl.java [hunk 6]
```diff
@@ -31,6 +31,9 @@
   @Override
   public GitFullSyncConfigDTO createConfig(
       String accountIdentifier, String orgIdentifier, String projectIdentifier, GitFullSyncConfigRequestDTO dto) {
+    // Why are we making a GET call to check the config?
+    // We can simply make a save call and it will automatically fail for duplicateKeyException if config
+    // already exists
     if (get(accountIdentifier, orgIdentifier, projectIdentifier).isPresent()) {
       throw new InvalidRequestException(
           "Configuration already exists, please update it instead of creating a new one", WingsException.USER);
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/core/fullsync/GitFullSyncConfigServiceImpl.java:39]: @mohitg0795 I feel this should also be improved.
"Configuration already exists. Update this configuration or create a new one."

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java [hunk 7]
```diff
@@ -219,15 +244,20 @@
...
           NGCommonEntityConstants.ACCOUNT_KEY) @io.harness.accesscontrol.AccountIdentifier String accountIdentifier,
       @Parameter(description = ORG_PARAM_MESSAGE) @QueryParam(
           NGCommonEntityConstants.ORG_KEY) @io.harness.accesscontrol.OrgIdentifier @OrgIdentifier String orgIdentifier,
       @Parameter(description = PROJECT_PARAM_MESSAGE) @QueryParam(NGCommonEntityConstants.PROJECT_KEY) @io.
       harness.accesscontrol.ProjectIdentifier @ProjectIdentifier String projectIdentifier,
+      // What is a search term? Do our customers understand it?
       @Parameter(description = GitSyncApiConstants.SEARCH_TERM_PARAM_MESSAGE) @QueryParam(
           NGResourceFilterConstants.SEARCH_TERM_KEY) String searchTerm,
+      // Why is it a required field?
       @RequestBody(description = "Filters like entityType and syncStatus",
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java:261]: @mohitg0795 How about "Filters that must be used for full sync of the given entity."?

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java [hunk 8]
```diff
@@ -219,15 +244,20 @@
   @NGAccessControlCheck(resourceType = ResourceTypes.PROJECT, permission = VIEW_PROJECT_PERMISSION)
   public ResponseDTO<PageResponse<GitFullSyncEntityInfoDTO>>
   listFiles(
+      // Please maintain order of parameters in any request
+      // Page request can't be the first parameter.
       @RequestBody(description = "Details of Page including: size, index, sort") @BeanParam PageRequest pageRequest,
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java:249]: Details to fetch paginated results.

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java [hunk 9]
```diff
@@ -200,16 +218,23 @@
...
+    // Also, we should return the config that's deleted. Simply returning true/false is sort of not a good response.
     return ResponseDTO.newResponse(
         gitFullSyncConfigService.delete(accountIdentifier, orgIdentifier, projectIdentifier));
   }
 
+  // Lets discuss on why its a blocker to make it a GET request and how can we try to resolve it
+  // Using POST for GET requests isn't correct
   @POST
   @Path("/files")
   @ApiOperation(value = "List files in full sync along with their status", nickname = "listFullSyncFiles")
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java:240]: Fetches the list of files and their status in full sync.
Also, isn't this a GET operation? Why have we used a POST?
cc: @mohitg0795

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java [hunk 10]
```diff
@@ -175,22 +187,28 @@
...
+                // Message can be better : "Git Full Sync Config not found at the scope provided" ?
                 () -> new InvalidRequestException("Config not found at the scope provided", WingsException.USER));
     return ResponseDTO.newResponse(gitFullSyncConfigDTO);
   }
 
   @DELETE
   @Path("/config")
   @ApiOperation(value = "Delete full sync configuration", nickname = "deleteGitFullSyncConfig")
+  // Summary should be : "Delete configuration used for performing git full sync" ?
   @Operation(operationId = "deleteGitFullSyncConfig", summary = "Delete configuration for performing git full sync",
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java:212]: Deletes full sync configuration for the given Full Sync ID.
I guess we maintain ID for full sync and must provide that for deletion?
cc: @mohitg0795

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java [hunk 11]
```diff
@@ -166,6 +177,7 @@
   @GET
   @Path("/config")
   @ApiOperation(value = "Get full sync configuration", nickname = "getGitFullSyncConfig")
+  // Summary should be : "Get configuration used for performing git full sync" ?
   @Operation(operationId = "getGitFullSyncConfig", summary = "Get configuration for performing git full sync",
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java:181]: Fetches configuration details to perform Git full sync.

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java [hunk 12]
```diff
@@ -143,15 +151,18 @@
   @PUT
   @Path("/config")
   @ApiOperation(value = "Update a full sync configuration", nickname = "updateGitFullSyncConfig")
+  // Summary should be : "Updates configuration used for performing git full sync" ?
   @Operation(operationId = "updateGitFullSyncConfig", summary = "Updates a configuration for performing git full sync",
       responses =
       {
+        // It should say : "Returns the `updated` git full sync configuration back along with scope information"
         @io.swagger.v3.oas.annotations.responses.
         ApiResponse(description = "Returns the configuration back along with the scope information")
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java:160]: Returns the updated configuration details and scope for the Git full sync.

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java [hunk 13]
```diff
@@ -143,15 +151,18 @@
   @PUT
   @Path("/config")
   @ApiOperation(value = "Update a full sync configuration", nickname = "updateGitFullSyncConfig")
+  // Summary should be : "Updates configuration used for performing git full sync" ?
   @Operation(operationId = "updateGitFullSyncConfig", summary = "Updates a configuration for performing git full sync",
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/remote/GitFullSyncResource.java:155]: Updates configuration for performing Git full sync for the given (?) ID.

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/dtos/TriggerFullSyncResponseDTO.java [hunk 14]
```diff
@@ -22,5 +22,6 @@
 @Schema(name = "TriggerFullSyncResponse", description = "This has details to trigger Full Sync")
 @OwnedBy(DX)
 public class TriggerFullSyncResponseDTO {
+  // File to review
   @Schema(description = "Indicates whether Full Sync is triggered or not") Boolean isFullSyncTriggered;
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/dtos/TriggerFullSyncResponseDTO.java:26]: Determines if full sync is triggered.

File: 136-git-sync-manager/src/main/java/io/harness/gitsync/common/dtos/TriggerFullSyncResponseDTO.java [hunk 15]
```diff
@@ -22,5 +22,6 @@
 @Schema(name = "TriggerFullSyncResponse", description = "This has details to trigger Full Sync")
```
Comment (rashminandasahoo) [inline 136-git-sync-manager/src/main/java/io/harness/gitsync/common/dtos/TriggerFullSyncResponseDTO.java:22]: Details to trigger full sync.

**Review by rashminandasahoo — CHANGES_REQUESTED** [review]

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Passed_** 
 


            
 No issues Everything looks good
 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/30411)

**Pull request closed without merging**